sklearn/impute.py

```diff
@@ -1265,3 +1255,283 @@ def fit_transform(self, X, y=None):
+            distance matrix between the receivers and the train samples
+
+        receivers_idx : array-like, shape = (n_receivers, )
+            indices of X_col to be imputed
+
+        X_col : array-like, shape = (n_samples, )
+            column to be imputed
+
+        fit_X_col : array-like, shape = (n_train_samples, )
+            column from training
+
+        mask_fit_X_col : array-like, shape = (n_train_samples, )
+            mask for missing values corresponding to ``fit_X_col``
+
+        statistic : float
+            statistic for column generated from fitting ``fit_X_col``
+        """
+        if receivers_idx.shape[0] == 0:
+            return
+
+        # Row index for receivers and potential donors (pdonors)
+        potential_donors_idx = np.where(~mask_fit_X_col)[0]
+
+        # Impute using column mean if n_neighbors are not available
+        if len(potential_donors_idx) < self.n_neighbors:
+            X_col[receivers_idx] = statistic
+            return
+
+        # Get distance from potential donors
+        dist_potential_donors = dist[:, potential_donors_idx]
+
+        # Get donors
+        donors_idx = np.argpartition(dist_potential_donors,
+                                     self.n_neighbors - 1,
+                                     axis=1)[:, :self.n_neighbors]
+
+        # Get weights or None
+        donors_dist = dist_potential_donors[
+            np.arange(donors_idx.shape[0])[:, None], donors_idx]
+        weight_matrix = _get_weights(donors_dist, self.weights)
+
+        # Retrieve donor values and calculate kNN score
+        donors = fit_X_col[potential_donors_idx].take(donors_idx)
+        donors = np.ma.array(donors,
+                             mask=_get_missing_mask(donors,
+                                                    self.missing_values))
+
+        # Final imputation
+        imputed = np.ma.average(donors, axis=1, weights=weight_matrix)
+        X_col[receivers_idx] = imputed.data
+
+    def fit(self, X, y=None):
+        """Fit the imputer on X.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features, )
+            Input data, where ``n_samples`` is the number of samples and
+            ``n_features`` is the number of features.
+
+        Returns
+        -------
+        self : object
+        """
+
+        # Check data integrity and calling arguments
+        if not is_scalar_nan(self.missing_values):
+            force_all_finite = True
+        else:
+            force_all_finite = "allow-nan"
+            if self.metric not in _NAN_METRICS and not callable(self.metric):
+                raise ValueError(
+                    "The selected metric does not support NaN values")
+        X = check_array(X, accept_sparse=False, dtype=FLOAT_DTYPES,
+                        force_all_finite=force_all_finite, copy=self.copy)
+
+        # Check if sufficient neighboring samples available
+        if X.shape[0] < self.n_neighbors:
```

Comment:
Maybe we do not need this check. If there are 5 observations and k=5, this imputer becomes a mean imputer. It still works, but it will not use any of the KNN concepts.